flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/datagen/SequenceGenerator.java

```diff
@@ -65,33 +71,46 @@ public void open(
                 this.checkpointedState == null,
                 "The " + getClass().getSimpleName() + " has already been initialized.");
 
-        this.checkpointedState =
-                context.getOperatorStateStore()
-                        .getListState(
-                                new ListStateDescriptor<>(
-                                        name + "-sequence-state", LongSerializer.INSTANCE));
-        this.valuesToEmit = new ArrayDeque<>();
-        if (context.isRestored()) {
-            // upon restoring
+        ListStateDescriptor<InternalState> stateDescriptor =
+                new ListStateDescriptor<>(
+                        name + "-sequence-state", TypeInformation.of(InternalState.class));
+        this.checkpointedState = context.getOperatorStateStore().getListState(stateDescriptor);
+        this.internalStates = Lists.newArrayList();
 
-            for (Long v : this.checkpointedState.get()) {
-                this.valuesToEmit.add(v);
-            }
+        if (context.isRestored()) {
+            checkpointedState.get().forEach(state -> internalStates.add(state));
         } else {
-            // the first time the job is executed
-            final int stepSize = runtimeContext.getNumberOfParallelSubtasks();
+            // The first time the job is executed.
             final int taskIdx = runtimeContext.getIndexOfThisSubtask();
-            final long congruence = start + taskIdx;
-
-            long totalNoOfElements = Math.abs(end - start + 1);
-            final int baseSize = safeDivide(totalNoOfElements, stepSize);
-            final int toCollect =
-                    (totalNoOfElements % stepSize > taskIdx) ? baseSize + 1 : baseSize;
+            final long stepSize = runtimeContext.getNumberOfParallelSubtasks();
+            InternalState state = new InternalState(taskIdx, stepSize, start + taskIdx);
+            internalStates.add(state);
+        }
+    }
 
-            for (long collected = 0; collected < toCollect; collected++) {
-                this.valuesToEmit.add(collected * stepSize + congruence);
+    public Long nextValue() {
+        Iterator<InternalState> iterator = internalStates.iterator();
```

Comment:
When do we have the case that there is more than one state in the `internalStates`? Is that when we're reducing the parallelism and multiple states get picked up by a single operator? Is it enough to always select the first `InternalState`? It feels to me that we're processing one state first until we reach the end and would then process the next state which has a way lower number again? :thinking:

Reply:
Yes, multiple states will only appear when the degree of parallelism is reduced.

I guess what you're thinking is: are we going to get the latest sequence value by putting the state into the queue and then taking it off the head of the queue each time and putting it at the end? Instead of sending all the sequences of a state and then sending the sequence value of the next state?

Reply:
I'll try to put it in unit tests... ..

Reply:
This seems like a good idea, and it's probably appropriate to create some JIRAs to implement in subsequent, incremental steps.

The current internal states (before this PR) for a generator from 1 to 1M would have 1 million values: `1,2,3...1000000`.  If it were checkpointed, it would **(1)** take quite a bit of space and **(2)** I'm not sure that restoring it to a lower parallelism retains any lowest-first semantics.

This PR solves **(1)**, but doesn't change much in terms of **(2)**.  The internal states would be `(1,3), (2,3), (3,3)` (assuming a parallelism of 3).  

If it was restored to a parallelism of two (after the PR), and one task got `(1,3), (2,3)` as its internal states, it would emit `1,4,7,10 ... 1000000, 2, 5, 8 ...`.  

If my understanding is correct, emitting `2` _after_ `1000000` is a possibility in both scenarios.  There's an opportunity here:

* Improve the logic so that a task always emits the lowest of the possible values from the internal states that it possesses (although a priority queue is probably unnecessary, it should end up roughly equivalent to sorting and cycling through the states).  This is really a new improvement that could be applied to the current state.
* Avoid performance skew after reducing parallelism.  Currently, going from N tasks to N-1 tasks will double the length of time it takes, since one task will have twice as many values to emit as the others.  This is a undesirable consequence of this PR, but can probably also be discussed and implemented in a subsequent step.

One thing that I really like about this PR is that it enables huge sequences that aren't possible today, in an efficient manner.  This is a really compelling advantage over the current state in every way!

Reply:
> * Avoid performance skew after reducing parallelism.  Currently, going from N tasks to N-1 tasks will double the length of time it takes, since one task will have twice as many values to emit as the others.  This is a undesirable consequence of this PR, but can probably also be discussed and implemented in a subsequent step.

Yes, I noticed this problem, but I haven't thought of a better solution.

But I have an idea, that is, for each subtask, we let him generate more InternalState, and the general process is described as follows:

Assuming that the parallelism is 5 and sending 100 pieces of data, we let each subtaks be logically divided into multiple InternalStates
The current InternalState is as follows
```java
InternalState{
         long stepSize;
         long nextValue;
}
```
When subtask-0 is initialized: only one InternalState(5,0) will be initialized;


So my idea is to add an end to indicate the end value of the current InternalState
```java
InternalState{
         long stepSize;
         long nextValue;
         long end;
}
```
Cutting when subtask-0 is initialized produces two: InternalState(5,0,50), InternalState(5,0,100);
Of course, it doesn’t have to be two, it can be multiple, so that when the parallelism is readjusted, one task can avoid processing multiple InternalStates.
@XComp @RyanSkraba  WDYT?

At that time, I will raise a jira to track this issue, and then we will discuss this issue.

Reply:
Thanks for sharing your view, @RyanSkraba. I think you're right - the implementation in master (i.e. without this PR's change) doesn't support reducing the parallelism properly, too. 
```
Avoid performance skew after reducing parallelism. Currently, going from N tasks to N-1 tasks will double the length of time it takes, since one task will have twice as many values to emit as the others. This is a undesirable consequence of this PR, but can probably also be discussed and implemented in a subsequent step.
```
I don't understand that part: Isn't the performance skew issue you're describing more general and not rely connected to the `DataGenerator`? The skew you're mentioning is happening in any case when reducing the parallelism. It's more of a question how Flink would distribute the operator state after the parallelism was reduced. :thinking: Therefore, I don't see this as a problem for this PR.

Reply:
@xuzhiwen1255 This sounds perfect!
> But I have an idea, that is, for each subtask, we let him generate more InternalState

That is to say, instead of having N internal states when there are N tasks initially, we would have 2N, or 10N, or 100N.  That number doesn't have to be extremely high to avoid skew here, and still benefit from the "concise" and efficient InternalState you've implemented here.

I think there might be some cool way of splitting up the ranges differently, but I'd like to think about it!  This should probably go on the future JIRA though :D 

@XComp Yes, this issue would happen to any source that created "one split per task" on initialization, where the split corresponds to the operator state -- but the previous implementation didn't do that.  It was one state per element to emit, so it _could_ be perfectly rebalanced (but overall pretty inefficiently).

I agree that it's not a problem for this PR!

Reply:
> That is to say, instead of having N internal states when there are N tasks initially, we would have 2N, or 10N, or 100N. That number doesn't have to be extremely high to avoid skew here, and still benefit from the "concise" and efficient InternalState you've implemented here.

Yes, but I haven't figured out how to solve this problem reasonably.